lib/charms/layer/kubernetes_common.py

```diff
@@ -933,3 +934,199 @@ def get_secret_password(username):
+        subprocess.check_call(["mount", "-a"])
+    else:
+        hookenv.log("LXD not detected, not faking kernel params")
+
+
+def get_sandbox_image_uri(registry):
+    return "{}/pause:3.4.1".format(registry)
+
+
+def configure_kubelet(dns_domain, dns_ip, registry, taints=None):
+    kubelet_opts = {}
+    kubelet_opts["kubeconfig"] = kubelet_kubeconfig_path
+    kubelet_opts["network-plugin"] = "cni"
+    kubelet_opts["v"] = "0"
+    kubelet_opts["logtostderr"] = "true"
+    kubelet_opts["node-ip"] = get_node_ip()
+
+    container_runtime = endpoint_from_flag("endpoint.container-runtime.available")
+
+    kubelet_opts["container-runtime"] = container_runtime.get_runtime()
+    if kubelet_opts["container-runtime"] == "remote":
+        kubelet_opts["container-runtime-endpoint"] = container_runtime.get_socket()
+
+    feature_gates = {}
+
+    kubelet_cloud_config_path = cloud_config_path("kubelet")
+    if is_state("endpoint.aws.ready"):
+        kubelet_opts["cloud-provider"] = "aws"
+        feature_gates["CSIMigrationAWS"] = False
+    elif is_state("endpoint.gcp.ready"):
+        kubelet_opts["cloud-provider"] = "gce"
+        kubelet_opts["cloud-config"] = str(kubelet_cloud_config_path)
+        feature_gates["CSIMigrationGCE"] = False
+    elif is_state("endpoint.openstack.ready"):
+        kubelet_opts["cloud-provider"] = "external"
+    elif is_state("endpoint.vsphere.joined"):
+        # vsphere just needs to be joined on the worker (vs 'ready')
+        kubelet_opts["cloud-provider"] = "vsphere"
+        # NB: vsphere maps node product-id to its uuid (no config file needed).
+        uuid = _get_vmware_uuid()
+        kubelet_opts["provider-id"] = "vsphere://{}".format(uuid)
+    elif is_state("endpoint.azure.ready"):
+        azure = endpoint_from_flag("endpoint.azure.ready")
+        kubelet_opts["cloud-provider"] = "azure"
+        kubelet_opts["cloud-config"] = str(kubelet_cloud_config_path)
+        kubelet_opts["provider-id"] = azure.vm_id
+        feature_gates["CSIMigrationAzureDisk"] = False
+
+    # Put together the KubeletConfiguration data
+    kubelet_config = {
+        "apiVersion": "kubelet.config.k8s.io/v1beta1",
+        "kind": "KubeletConfiguration",
+        "address": "0.0.0.0",
+        "authentication": {
+            "anonymous": {"enabled": False},
+            "x509": {"clientCAFile": str(ca_crt_path)},
+        },
+        # NB: authz webhook config tells the kubelet to ask the api server
+        # if a request is authorized; it is not related to the authn
+        # webhook config of the k8s master services.
+        "authorization": {"mode": "Webhook"},
+        "clusterDomain": dns_domain,
+        "failSwapOn": False,
+        "port": 10250,
+        "protectKernelDefaults": True,
+        "readOnlyPort": 0,
+        "tlsCertFile": str(server_crt_path),
+        "tlsPrivateKeyFile": str(server_key_path),
+    }
+    if dns_ip:
+        kubelet_config["clusterDNS"] = [dns_ip]
```

Comment:
This could use a second look.

currently in kubernetes-worker if this is the case

dns['sdn-ip'] = <something>
dns['enable-kube-dns'] = False
kubelet_config["clusterDNS"] would be unset.

After these changes:
kubelet_config["clusterDNS"] = [dns_ip]

is this ok?
dns comes from interface-kube-control

Reply:
is is possible the changes in [`kubernetes-master`](https://github.com/charmed-kubernetes/charm-kubernetes-master/pull/191/files#diff-b3099d2217027f2f460740149252f0dca742dbb5deeb274bc17559fb648b1d86L1289-R1328)
are intended to prevent this situation?

If that is true, does this mean worker charm must be upgraded after the master?

Reply:
This should be okay, since enable-kube-dns and sdn-ip are tightly coupled.

Before changes, kubernetes-master sets sdn-ip to None whenever enable-kube-dns is False ([link](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/2d7eda74e22b1fe67e2ed0ae17556c95da935077/reactive/kubernetes_master.py#L1289)). Likewise, it sets enable-kube-dns to True whenever sdn-ip is not None ([link](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/2d7eda74e22b1fe67e2ed0ae17556c95da935077/reactive/kubernetes_master.py#L1294-L1296), [link](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/2d7eda74e22b1fe67e2ed0ae17556c95da935077/reactive/kubernetes_master.py#L1310)).

After changes, kubernetes-master sets enable-kube-dns explicitly to `dns_ip is not None` ([link](https://github.com/charmed-kubernetes/charm-kubernetes-master/pull/191/files#diff-b3099d2217027f2f460740149252f0dca742dbb5deeb274bc17559fb648b1d86R1345)).

Behavior looks the same before and after so I don't think workers would need to be upgraded after the master.

I did it this way to avoid adding a `dns_enabled` argument to configure_kubelet, since that struck me as redundant. But maybe doing it this way is more confusing than it is clean. What do you think?

Reply:
I appreciate the `dns_ip = None` meaning not enabled.
i think my worry was that because of the [relation data](https://github.com/juju-solutions/interface-kube-control/blob/534310f5bca8edde02cadaf6ac42231cea0b040b/requires.py#L75-L86) that there might be a race condition where the worker charm is updated before the master -- and leaving the master charm in this state:

```python
dns['sdn-ip'] = <something>
dns['enable-kube-dns'] = False
```

I could be TOTALLY reading to deep in a situation that isn't relevant

Reply:
Yup, like I said before, kubernetes-master sets sdn-ip to None whenever it sets enable-kube-dns to False and that was true before I made any changes to it. So I think we're good.